docs/specs/lcb-traits.md

```diff
@@ -0,0 +1,364 @@
+*and can be verified* at compile time.  This doesn't satisfy the requirement.
+
+For example, it would be convenient to write a `Log()` handler that can log
+any type that can also be formatted.  However, with the generic handlers scheme
+described above, `Log()` would have to take arguments as `any` and defer
+type-checking to runtime.
+
+## Traits
+
+Traits are collections of methods that can be used to extend the functionality
+of a type.
+
+Traits have no storage associated with them.  Storage is provided by
+implementing the trait for a specific storage type.
+
+### Defining traits
+
+    public trait type CanFormatAsString
+        method type FormatAsString(in me) as string
+    end type
+
+A trait definition is a block containing a list of zero or more `method type`
+declarations.  `method type` declarations take the same form as existing
+`handler type` declarations.
+
+Two additional symbols are defined within a trait definition:
+
+1. `MyType` is always equivalent to the type of the value that the method is
+   being called for.
+2. `me` must always be the first parameter a `method` definition or
+   `method type` declaration; it is equivalent to writing `me as MyType`.  The
+   `me` parameter must always be `in` or `inout`.
+
+### Implementing traits
+
+    public implementation of CanFormatAsString for T
+        method FormatAsString(in me) as string
+            variable tFormatted as string
+            -- Format me parameter
+            return tFormatted
+        end method
+    end implementation
+
+`method` definitions take the same form as existing `handler` definitions.
+
+In order for an `implementation of U for T` to be valid, the following criteria
+must be fulfilled:
+
+1. **Completeness Rule**.  All methods of `U` must be implemented (either in
+   the `implementation` block or via a default implementation in the trait
+   definition).
+2. **Locality Rule**.  Either `U` or `T` must be local, i.e. defined in the
+   current module.
+3. **Visibility Rule**.  The visibility of the implementation must be at least
+   the accessibility of `U` (i.e. if `U` is `public`, then the implementation
+   must also be `public`).
+
+#### Completeness rule
+
+The **Completeness Rule** is required to ensure that it is possible to call any
+method of a trait `U` for any value of type `T` for which `U` has been
+implemented.
+
+#### Locality rule
+
+The **Locality Rule** prevents two implementations of `T`.  Without this rule,
+the following scenario would be possible:
+
+* Module `A` defines and type `T`
+* Module `B` implements `CanFormatAsString` for `T`
+* Module `C` implements `CanFormatAsString` for `T` (differently)
+* Module `D` uses modules `B` and `C` and attempts to format an instance of `T`
+  as a string.  Which implementation should be used?
+
+By requiring either the trait type or the concrete type to be defined locally
+to the implementation, this diamond problem can never occur.
+
```

Comment:
Yes - for a 'derived' type (which we still need to define exactly what that means in terms of traits) a BobT 'will be' an AliceT (in some sense). However, the use-case is using a third-party type and a third-party trait and joining them by your implementation of the trait for the type... This means that the values of type AliceT will be coming from somewhere outwith your module - and thus will be AliceT's and not BobT's. Thus, if you want your trait to work whenever you get an AliceT you need to 'do something' to make it a BobT. (It would probably have to be aggregation rather than derivation).